src/segment-loader.js

```diff
@@ -1260,20 +1264,44 @@ export default class SegmentLoader extends videojs.EventTarget {
     // can be adjusted by the timestamp offset
     if (!segmentInfo.hasAppendedData_) {
       this.metadataQueue_.caption.push(
-        this.handleCaptions_.bind(this, simpleSegment, captions));
+        this.handleCaptions_.bind(this, simpleSegment, captionData));
       return;
     }
 
     const timestampOffset = this.sourceUpdater_.videoTimestampOffset() === null ?
       this.sourceUpdater_.audioTimestampOffset() :
       this.sourceUpdater_.videoTimestampOffset();
 
-    captions.forEach((caption) => {
-      createCaptionsTrackIfNotExists(
-        this.inbandTextTracks_, this.hls_.tech_, caption.stream);
+    const tracks = {};
+
+    // get total start/end and captions for each track/stream
+    captionData.forEach((caption) => {
+      // caption.stream is actually a track name...
+      const track = tracks[caption.stream] = tracks[caption.stream] || {
+        startTime: Infinity,
+        captions: [],
+        endTime: 0
+      };
+
+      track.startTime = Math.min(track.startTime, (caption.startTime + timestampOffset));
+      track.endTime = Math.max(track.endTime, (caption.endTime + timestampOffset));
+      track.captions.push(caption);
+    });
+    Object.keys(tracks).forEach((trackName) => {
+      const {startTime, endTime, captions} = tracks[trackName];
+      const {inbandTextTracks_} = this;
+
+      this.logger_(`adding cues from ${startTime} -> ${endTime} for ${trackName}`);
+
+      createCaptionsTrackIfNotExists(inbandTextTracks_, this.hls_.tech_, trackName);
+      // clear out any old cues. We do this because a rendition change
+      // will cause captions to be re-parsed for certain segments, as they are
+      // downloaded, but we do not want to have two of the same captions visible.
+      removeCuesFromTrack(startTime, endTime, inbandTextTracks_[trackName]);
```

Comment:
This is the actual fix, but there was a lot of needed refactor to get here.

Reply:
won't this end up clearing out all the cues except the most recently parsed caption cue if you don't rendition switch?

Reply:
This clears out all cues in the range of time that the current cues are going to be inserted. IE: If we have cues from 10-20s and we get more cues from 10-20s we clear the previous cues from 10-20s before inserting the new ones. From what I have seen this only happens during a rendition switch with a timescale change.